zilliqa/src/precompiles.rs

```diff
@@ -63,14 +69,85 @@ impl ERC20Precompile {
     }
 }
 
-fn make_selector(signature: &str) -> [u8; 4] {
-    let signature = Keccak256::digest(signature).to_vec();
-    let slice = signature.as_slice();
-    let res: [u8; 4] = slice[..4].try_into().unwrap();
-    res
+impl ContextStatefulPrecompile<&State> for ERC20Precompile {
+    fn call(
+        &self,
+        input: &Bytes,
+        gas_price: u64,
+        context: &mut InnerEvmContext<&State>,
+    ) -> PrecompileResult {
+        if input.length() < 4 {
+            return Err(PrecompileError::Other(
+                "Provided input must be at least 4-byte long".into(),
+            )
+            .into());
+        }
+
+        let dispatch_table: [([u8; 4], _); 1] = [(
+            short_signature("balanceOf", &[ParamType::Address]),
+            Self::get_balance,
+        )];
+
+        let Some(handler) = dispatch_table
+            .iter()
+            .find(|&predicate| predicate.0 == input[..4])
+        else {
+            return Err(PrecompileError::Other(
+                "Unable to find handler with given selector".to_string(),
+            )
+            .into());
+        };
+
+        handler.1(&input[4..], gas_price, context)
+    }
 }
 
-impl ContextStatefulPrecompile<&State> for ERC20Precompile {
+pub(crate) struct PopVerifyPrecompile;
+
+// keep in-sync with zilliqa/src/contracts/deposit.sol
+impl PopVerifyPrecompile {
+    const POP_VERIFY_GAS_PRICE: u64 = 1_000_000u64; // FIXME: Gas Price?
```

Comment:
I'm not sure what is a good number to use as the gas price. So, I arbitrarily put in 1M.

Reply:
Happy to leave like this for now and discuss with @DrZoltanFazekas soon :)